File: tests/google_search.py
results = None
result = None
href = None


def init_results(driver):
    global results, result
    driver.wait_for_load()
    results = driver.find_elements('css selector', '#search .r > a')
    if results:
        result = results[1]
        return True
    else:
        return False


def assert_link_unchanged():
    global result, href
    assert result.get_attribute('href') == href


def assert_no_intermediate_urls(driver, method, target):
    driver.get_urls()
    method()
    urls = []

    def check_urls():
        urls.extend(driver.get_urls())
        return urls
    driver.wait_until(check_urls)
    assert urls[0] == target


def run(driver):
    global href

    # Search for site:palant.de
    driver.navigate('https://www.google.com/?gfe_rd=cr&hl=en')
    driver.wait_until(lambda: driver.find_elements('name', 'q'))
    driver.find_element('name', 'q').send_keys('site:palant.de', driver.keys.RETURN)

    # Choose a search result
    driver.wait_until(lambda: init_results(driver))
    href = result.get_attribute('href')
    assert 'google.com' not in href

    # Right-click the search result
    result.context_click()
    assert_link_unchanged()
    result.send_keys(driver.keys.ESCAPE)
    assert_link_unchanged()

    # Click the search result
    with driver.restore_url():
        assert_no_intermediate_urls(driver, lambda: result.click(), href)
    driver.wait_until(lambda: init_results(driver))
    assert_link_unchanged()

    # Press Enter on a search result
    with driver.restore_url():
        assert_no_intermediate_urls(driver,
                                    lambda: result.send_keys(driver.keys.RETURN),
                                    href)
    driver.wait_until(lambda: init_results(driver))
    assert_link_unchanged()

    # Middle-click search result
    assert_no_intermediate_urls(driver, lambda: result.middle_click(), href)
    driver.close_background_tabs()
    assert_link_unchanged()

File: test/google_search.js
/*
 * This Source Code is subject to the terms of the Mozilla Public License
 * version 2.0 (the "License"). You can obtain a copy of the License at
 * http://mozilla.org/MPL/2.0/.
 */

"use strict";

import * as utils from "../test-utils.js";

const searchFieldSelector = "[name=q]";
const searchResultSelector = "#search .r > a";

describe("Google Search", () =>
{
  let browser;
  let page;
  let href;

  it("should load", async function()
  {
    browser = await utils.launchBrowser();
    page = await browser.newPage();
    await page.goto("https://www.google.com/?gfe_rd=cr&hl=en");
  });

  it("should allow searching", async function()
  {
    let searchField = await page.$(searchFieldSelector);
    expect(searchField).to.be.not.null;

    await searchField.type("site:palant.de\n");
    await page.waitForSelector(searchResultSelector);
  });

  it("should not link to Google in results", async function()
  {
    let result = await page.$(searchResultSelector);
    expect(result).to.be.not.null;

    href = await result.getHref();
    expect(href).to.not.include("google.com");
  });


  it("shouldn't change link on hover", async function()
  {
    let result = await page.$(searchResultSelector);

    await result.hover();
    await page.waitFor(200);
    expect(await result.getHref()).to.equal(href);
  });

  it("shouldn't change link on right-click", async function()
  {
    let result = await page.$(searchResultSelector);

    await result.click({button: "right"});
    await page.waitFor(200);
    expect(await result.getHref()).to.be.equal(href);
  });

  it("shouldn't change link on middle-click", async function()
  {
    let result = await page.$(searchResultSelector);

    let target = await utils.newTarget(browser, async function()
    {
      await result.click({button: "middle"});
    });
    expect(target.url()).to.be.equal(href);

    let tab = await target.page();
    await tab.close();

    await page.waitFor(200);
    expect(await result.getHref()).to.be.equal(href);
  });

  it("shouldn't navigate to search result via intermediate URLs when clicked", async function()
  {
    let result = await page.$(searchResultSelector);

    let requests = await utils.logTopLevelRequests(page, async function()
    {
      await result.click();
      await page.waitForNavigation();
    });

    try
    {
      expect(await page.url()).to.be.equal(href);
      expect(requests).to.be.deep.equal([href]);
    }
    finally
    {
      await page.goBack();
    }
  });

  it("shouldn't navigate to search result via intermediate URLs when Enter is pressed", async function()
  {
    let result = await page.$(searchResultSelector);

    let requests = await utils.logTopLevelRequests(page, async function()
    {
      await result.press("Enter");
      await page.waitForNavigation();
    });

    try
    {
      expect(await page.url()).to.be.equal(href);
      expect(requests).to.be.deep.equal([href]);
    }
    finally
    {
      await page.goBack();
    }
  });

  it("should shut down", async function()
  {
    await browser.close();
  });
});
